Add HeaderDoc comments to the Web Extension headers.

xeenon opened the pull request:
#### 1e49c3463baf4ccdb7f56120c1af54d4564918df
<pre>
Add HeaderDoc comments to the Web Extension headers.
<a href="https://bugs.webkit.org/show_bug.cgi?id=247844">https://bugs.webkit.org/show_bug.cgi?id=247844</a>

Reviewed by Brian Weinstein.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext setPermissionState:forURL:]): Drive-by. Add missing assert.
(-[_WKWebExtensionContext setPermissionState:forURL:expirationDate:]): Ditto.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMatchPattern.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPermission.h:

Canonical link: <a href="https://commits.webkit.org/256746@main">https://commits.webkit.org/256746@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f4ddff710c8f4eda785fe07c2d76509c67d75233

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96721 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5987 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/29815 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106249 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6200 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/34722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89106 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/102960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102397 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/83333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/31598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74528 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/25 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/22 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4807 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/83333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2242 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/1247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/40531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
```diff
@@ -52,52 +65,144 @@ typedef NS_ERROR_ENUM(_WKWebExtensionErrorDomain, _WKWebExtensionError) {
 - (instancetype)initWithAppExtensionBundle:(NSBundle *)appExtensionBundle NS_DESIGNATED_INITIALIZER;
+
+/*!
+ @abstract Returns a web extension initialized with a specified app extension bundle.
```
I think this should say `initialized with a specified resource base URL`.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
```diff
@@ -52,52 +65,144 @@ typedef NS_ERROR_ENUM(_WKWebExtensionErrorDomain, _WKWebExtensionError) {
+ @param manifestVersion The version number to check.
+ @result Returns `YES` if the extension specified a manifest version that is is greater than or equal to `manifestVersion`.
+ */
 - (BOOL)usesManifestVersion:(double)manifestVersion;
```
I missed this in earlier review, but is there a reason for the manifestVersion to be a double instead of an int?

xeenon replied:
Future proofing. There have been talks in http://github.com/w3c/webextensions about maybe doing a 3.1 or dot versions in general.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
```diff
@@ -52,52 +65,144 @@ typedef NS_ERROR_ENUM(_WKWebExtensionErrorDomain, _WKWebExtensionError) {
 
-@property (readonly, nonatomic) BOOL hasBackgroundContent;
-@property (readonly, nonatomic) BOOL backgroundContentIsPersistent;
+/*! @abstract The websites that the extension needs to access for base functionality, including injected content and websites that can send messaged to the extension. */
```
Typo: "that can send messaged"

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
```diff
@@ -36,126 +36,393 @@
-    _WKWebExtensionContextPermissionStateRequestedExplicitly = 1,
-    _WKWebExtensionContextPermissionStateGrantedImplicitly   = 2,
-    _WKWebExtensionContextPermissionStateGrantedExplicitly   = 3,
+    _WKWebExtensionContextPermissionStateUnknown             =  0,
```
Is this the suggested coding style?

xeenon replied:
WebKit does usually line up the numbers in enums.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
```diff
@@ -36,126 +36,393 @@
+/*! @abstract The extension controller this context is loaded in, otherwise `nil` if it isn't loaded. */
 @property (nonatomic, readonly, weak) _WKWebExtensionController *extensionController;
 
+/*! @abstract A boolean value indicating if this context is loaded in an extension controller. */
```
Super nit: I saw you used `Boolean` in some places and `boolean` here. Might be good to be consistent.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
```diff
@@ -36,126 +36,393 @@
-// When set to YES, this allows an extension to upgrade to asking for all URLs
-// without actually granting all URLs from an optional permissions request.
+/*!
+ @abstract A boolean value indicating if the extension has requested optioanl access to all hosts.
```
Typo: optioanl

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
```diff
@@ -36,126 +36,393 @@
+/*!
+ @abstract A boolean value indicating if the extension has requested optioanl access to all hosts.
+ @discussion When this value is `YES` the extension has asked for access to all hosts in a call to `browser.runtime.permissions.request()`
+ and future permission checks will present discrite hosts for approval as being implicty requested. This value should be saved and restored as needed.
```
Typo: discrite

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
```diff
@@ -36,126 +36,393 @@
+/*!
+ @abstract Checks if the currently granted permission match patterns set contains the `<all_urls>` pattern.
+ @discussion This does not check for any `*` host patterns. In most cases you should use the broader `hasAccessToAllHosts`.
+ This check is primarily needed for APIs like `browser.tabs.captureVisibleTab()` that need to check specifically for `<all_urls>`.
```
Are there any other APIs that use this check, or is it truly only for `captureVisibleTab`?

xeenon replied:
IIRC it is only `captureVisibleTab` but clients could use this for permission warning text as they see fit.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
```diff
@@ -36,126 +36,393 @@
+- (_WKWebExtensionContextPermissionState)permissionStateForURL:(NSURL *)url inTab:(nullable id <_WKWebExtensionTab>)tab NS_SWIFT_NAME(permissionState(forURL:in:));
 
+/*!
+ @abstract Sets the state of a URL in all tabs and global contexts with a distant furtue expiration date.
```
Typo: furtue

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
```diff
@@ -36,126 +36,393 @@
+/*!
+ @abstract Checks the specified match pattern against the currently denied, granted, and requested permission match patterns.
+ @discussion Match patterns can be granted on a per-tab basis. When the tab is known, access checks should always use the method that checks in a tab.
+ @seealso permissionStateForMatchPattern:inTab:
```
I think there was a copy-past-o here. This is `permissionStateForMatchPattern:inTab:`. This comment should probably mention this is the method that checks the tab.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
```diff
@@ -33,20 +33,69 @@
+/*!
+ @abstract Loads the specified extension context.
+ @discussion Causes the context to start, loading any background content, and injecting any content into relevant tabs.
+ @param error Set to \c nil or an \c NSError instance if an error occurred.
```
What is `\c`?

xeenon replied:
Shows it in code style in Xcode. Backticks don't work on the `@param` lines for some reason.

xeenon committed:
Add HeaderDoc comments to the Web Extension headers.
https://bugs.webkit.org/show_bug.cgi?id=247844

Reviewed by Brian Weinstein.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext setPermissionState:forURL:]): Drive-by. Add missing assert.
(-[_WKWebExtensionContext setPermissionState:forURL:expirationDate:]): Ditto.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMatchPattern.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPermission.h:

Canonical link: https://commits.webkit.org/256746@main
